File: csharp/src/Sigv4.cs
// AWS Signature Version 4, hand-rolled.
//
// The AWS providers need exactly one thing from the AWS SDK - request
// signing - and taking the SDK for it would break the no-dependency rule
// that keeps ten ports honest. SigV4 is a stable, published algorithm
// built from HMAC-SHA256, which the base class library already has.
//
// `Sign` is pure: the caller passes the timestamp, so the same input
// yields the same signature everywhere. That is what lets the shared spec
// carry known-answer cases that all ten ports must reproduce bit-for-bit,
// and lets the integration mock recompute the signature server-side.
//
// A port of typescript/src/Sigv4.ts, which is canonical.

using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

namespace Voxgig.Sekreto
{
    public static class Sigv4
    {
        /// <summary>
        /// Sign one request. Input keys: method, url, headers?, body?,
        /// service, region, keyid, secret, session?, and datetime
        /// (`YYYYMMDDTHHMMSSZ`; passed in, never sampled, so the function
        /// stays pure). Returns the headers to attach to the request:
        /// authorization, x-amz-date, and x-amz-security-token when a
        /// session token was given.
        /// </summary>
        public static Dictionary<string, object> Sign(Dictionary<string, object> input)
        {
            var url = new Uri(Text(input.GetValueOrDefault("url")));

            string datetime = Text(input.GetValueOrDefault("datetime"));
            string date = datetime.Substring(0, 8);
            string body = Text(input.GetValueOrDefault("body")) ?? "";

            string method = Text(input.GetValueOrDefault("method"));
            string service = Text(input.GetValueOrDefault("service"));
            string region = Text(input.GetValueOrDefault("region"));
            string keyid = Text(input.GetValueOrDefault("keyid"));
            string secret = Text(input.GetValueOrDefault("secret"));
            string session = Text(input.GetValueOrDefault("session"));

            // Every header that will be signed: the caller's extras, plus
            // host and x-amz-date (and the session token when present),
            // lower-cased and trimmed the way the canonical form requires.
            var headers = new Dictionary<string, string>();

            if (input.GetValueOrDefault("headers") is Dictionary<string, object> extra)
            {
                foreach (var entry in extra)
                {
                    headers[entry.Key.ToLowerInvariant()] = Convert.ToString(entry.Value).Trim();
                }
            }

            // Uri.Authority is the host plus any non-default port, exactly
            // the WHATWG `host` the canonical implementation signs.
            headers["host"] = url.Authority;
            headers["x-amz-date"] = datetime;
            if (!string.IsNullOrEmpty(session))
            {
                headers["x-amz-security-token"] = session;
            }

            var names = new List<string>(headers.Keys);
            names.Sort(StringComparer.Ordinal);

            var canonicalheaders = new StringBuilder();
            foreach (string name in names)
            {
                canonicalheaders.Append(name).Append(':').Append(headers[name]).Append('\n');
            }
            string signedheaders = string.Join(";", names);

            string path = url.AbsolutePath;
            if (0 == path.Length)
            {
                path = "/";
            }

            string canonicalrequest = string.Join("\n", new[]
            {
                method.ToUpperInvariant(),
                path,
                CanonicalQuery(url.Query.TrimStart('?')),
                canonicalheaders.ToString(),
                signedheaders,
                Sha256Hex(body),
            });

            string scope = date + "/" + region + "/" + service + "/aws4_request";

            string stringtosign = string.Join("\n", new[]
            {
                "AWS4-HMAC-SHA256",
                datetime,
                scope,
                Sha256Hex(canonicalrequest),
            });

            byte[] kdate = Hmac(Encoding.UTF8.GetBytes("AWS4" + secret), date);
            byte[] kregion = Hmac(kdate, region);
            byte[] kservice = Hmac(kregion, service);
            byte[] ksigning = Hmac(kservice, "aws4_request");
            string signature = Hex(Hmac(ksigning, stringtosign));

            var out_ = new Dictionary<string, object>
            {
                ["authorization"] = "AWS4-HMAC-SHA256 Credential=" + keyid + "/" + scope
                    + ", SignedHeaders=" + signedheaders + ", Signature=" + signature,
                ["x-amz-date"] = datetime,
            };

            if (!string.IsNullOrEmpty(session))
            {
                out_["x-amz-security-token"] = session;
            }

            return out_;
        }

        /// <summary>
        /// The canonical query string: each pair RFC 3986-escaped, sorted by
        /// escaped key then escaped value.
        /// </summary>
        private static string CanonicalQuery(string query)
        {
            if ("" == query)
            {
                return "";
            }

            var pairs = new List<string[]>();

            foreach (string pair in query.Split('&'))
            {
                int eq = pair.IndexOf('=');
                string key = -1 == eq ? pair : pair.Substring(0, eq);
                string value = -1 == eq ? "" : pair.Substring(eq + 1);
                pairs.Add(new[]
                {
                    UriEscape(Uri.UnescapeDataString(key)),
                    UriEscape(Uri.UnescapeDataString(value)),
                });
            }

            pairs.Sort((left, right) =>
            {
                int bykey = string.CompareOrdinal(left[0], right[0]);
                return 0 == bykey ? string.CompareOrdinal(left[1], right[1]) : bykey;
            });

            var out_ = new List<string>();
            foreach (string[] pair in pairs)
            {
                out_.Add(pair[0] + "=" + pair[1]);
            }

            return string.Join("&", out_);
        }

        /// <summary>
        /// RFC 3986 escaping, spelled out rather than borrowed: AWS wants
        /// every byte outside the unreserved set percent-encoded with
        /// uppercase hex, including `!`, `'`, `(`, `)` and `*`, and library
        /// encoders disagree on exactly those.
        /// </summary>
        private static string UriEscape(string text)
        {
            var out_ = new StringBuilder();

            foreach (byte head in Encoding.UTF8.GetBytes(text))
            {
                char ch = (char)head;

                if (('A' <= ch && 'Z' >= ch) || ('a' <= ch && 'z' >= ch)
                    || ('0' <= ch && '9' >= ch)
                    || '-' == ch || '_' == ch || '.' == ch || '~' == ch)
                {
                    out_.Append(ch);
                }
                else
                {
                    out_.Append('%').Append(((int)head).ToString("X2"));
                }
            }

            return out_.ToString();
        }

        private static string Sha256Hex(string text)
        {
            return Hex(SHA256.HashData(Encoding.UTF8.GetBytes(text)));
        }

        private static byte[] Hmac(byte[] key, string text)
        {
            using var mac = new HMACSHA256(key);
            return mac.ComputeHash(Encoding.UTF8.GetBytes(text));
        }

        private static string Hex(byte[] data)
        {
            var out_ = new StringBuilder(data.Length * 2);

            foreach (byte head in data)
            {
                out_.Append(head.ToString("x2"));
            }

            return out_.ToString();
        }

        private static string Text(object value)
        {
            return null == value ? null : Convert.ToString(value);
        }
    }
}

File: kotlin/src/Sigv4.kt
// AWS Signature Version 4, hand-rolled.
//
// The AWS providers need exactly one thing from the AWS SDK - request
// signing - and taking the SDK for it would break the no-dependency rule
// that keeps ten ports honest. SigV4 is a stable, published algorithm built
// from HMAC-SHA256, which the JDK already has.
//
// `sigv4` is pure: the caller passes the timestamp, so the same input
// yields the same signature everywhere. That is what lets the shared spec
// carry known-answer cases that all ports must reproduce bit-for-bit, and
// lets the integration mock recompute the signature server-side.
//
// A port of typescript/src/Sigv4.ts, which is canonical.

package com.voxgig.sekreto

import java.io.ByteArrayOutputStream
import java.net.URI
import java.nio.charset.StandardCharsets
import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.TreeMap
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * One request to sign - the same declarative shape the shared spec uses.
 * `datetime` is `YYYYMMDDTHHMMSSZ`, and it is the caller's, so that signing
 * is a pure function of its input.
 */
data class Signing(
    val method: String,
    val url: String,
    val service: String,
    val region: String,
    val keyid: String,
    val secret: String,
    val datetime: String,
    val headers: Map<String, String> = emptyMap(),
    val body: String = "",
    val session: String? = null,
)

internal fun hex(bytes: ByteArray): String {
    val out = StringBuilder()

    for (byte in bytes) {
        out.append("%02x".format(byte.toInt() and 0xff))
    }

    return out.toString()
}

internal fun sha256hex(text: String): String =
    try {
        hex(MessageDigest.getInstance("SHA-256").digest(text.toByteArray(StandardCharsets.UTF_8)))
    } catch (err: NoSuchAlgorithmException) {
        // Every JDK ships SHA-256; a JVM without it cannot sign anything.
        throw SekretoError("sekreto: sigv4: no SHA-256: ${err.message}")
    }

internal fun hmac(key: ByteArray, text: String): ByteArray =
    try {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        mac.doFinal(text.toByteArray(StandardCharsets.UTF_8))
    } catch (err: GeneralSecurityException) {
        throw SekretoError("sekreto: sigv4: no HmacSHA256: ${err.message}")
    }

/**
 * RFC 3986 escaping, which is stricter than the usual URL encoder: AWS
 * wants everything but unreserved characters escaped, with uppercase hex.
 */
internal fun uriescape(text: String): String {
    val out = StringBuilder()

    for (byte in text.toByteArray(StandardCharsets.UTF_8)) {
        val ch = byte.toInt() and 0xff
        if (('A'.code <= ch && 'Z'.code >= ch) ||
            ('a'.code <= ch && 'z'.code >= ch) ||
            ('0'.code <= ch && '9'.code >= ch) ||
            '-'.code == ch || '_'.code == ch || '.'.code == ch || '~'.code == ch
        ) {
            out.append(ch.toChar())
        } else {
            out.append('%').append("%02X".format(ch))
        }
    }

    return out.toString()
}

/** Percent-decode, and nothing else: `+` stays `+`, as on the wire. */
internal fun uridecode(text: String): String {
    val out = ByteArrayOutputStream()
    var index = 0

    while (index < text.length) {
        val head = text[index]

        if ('%' == head && index + 2 < text.length) {
            val code = text.substring(index + 1, index + 3).toIntOrNull(16)
            if (null != code) {
                out.write(code)
                index += 3
                continue
            }
            // A stray % is kept as-is, the way a browser would.
        }

        out.write(head.toString().toByteArray(StandardCharsets.UTF_8))
        index++
    }

    return String(out.toByteArray(), StandardCharsets.UTF_8)
}

/**
 * The canonical query string: each pair RFC 3986-escaped, sorted by escaped
 * key then escaped value.
 */
internal fun canonicalquery(query: String): String {
    if (query.isEmpty()) {
        return ""
    }

    return query.split("&")
        .map { pair ->
            val eq = pair.indexOf('=')
            val key = if (-1 == eq) pair else pair.substring(0, eq)
            val value = if (-1 == eq) "" else pair.substring(eq + 1)
            uriescape(uridecode(key)) to uriescape(uridecode(value))
        }
        .sortedWith(compareBy({ it.first }, { it.second }))
        .joinToString("&") { "${it.first}=${it.second}" }
}

/**
 * Sign one request. Returns the headers to attach: authorization,
 * x-amz-date, and x-amz-security-token when a session token was given, in
 * that order - the spec compares the result as a JSON object, and callers
 * print it field by field.
 */
fun sigv4(input: Signing): Map<String, String> {
    val url = URI.create(input.url)

    val date = input.datetime.substring(0, 8)
    val session = if (input.session.isNullOrEmpty()) null else input.session

    // Every header that will be signed: the caller's extras, plus host and
    // x-amz-date (and the session token when present), lower-cased and
    // trimmed the way the canonical form requires. A TreeMap keeps them
    // sorted by name, which is the canonical order.
    //
    // Canonical header values are trimmed AND internally collapsed - AWS
    // folds sequential whitespace to one space before signing, so a header
    // like "a  b" must sign as "a b" or the service refuses it.
    val headers = TreeMap<String, String>()
    for ((key, value) in input.headers) {
        headers[key.lowercase()] = value.trim().replace(Regex("\\s+"), " ")
    }
    headers["host"] = url.host + if (-1 == url.port) "" else ":${url.port}"
    headers["x-amz-date"] = input.datetime
    if (null != session) {
        headers["x-amz-security-token"] = session
    }

    val canonicalheaders = headers.entries.joinToString("") { "${it.key}:${it.value}\n" }
    val signedheaders = headers.keys.joinToString(";")

    val path = if (url.rawPath.isNullOrEmpty()) "/" else url.rawPath
    val query = url.rawQuery ?: ""

    val canonicalrequest = listOf(
        input.method.uppercase(),
        path,
        canonicalquery(query),
        canonicalheaders,
        signedheaders,
        sha256hex(input.body),
    ).joinToString("\n")

    val scope = "$date/${input.region}/${input.service}/aws4_request"

    val stringtosign = listOf(
        "AWS4-HMAC-SHA256",
        input.datetime,
        scope,
        sha256hex(canonicalrequest),
    ).joinToString("\n")

    val kdate = hmac(("AWS4" + input.secret).toByteArray(StandardCharsets.UTF_8), date)
    val kregion = hmac(kdate, input.region)
    val kservice = hmac(kregion, input.service)
    val ksigning = hmac(kservice, "aws4_request")
    val signature = hex(hmac(ksigning, stringtosign))

    val out = LinkedHashMap<String, String>()
    out["authorization"] =
        "AWS4-HMAC-SHA256 Credential=${input.keyid}/$scope" +
        ", SignedHeaders=$signedheaders" +
        ", Signature=$signature"
    out["x-amz-date"] = input.datetime

    if (null != session) {
        out["x-amz-security-token"] = session
    }

    return out
}
